src/plot_api/plot_api.js

```diff
@@ -2267,7 +2267,10 @@ exports.react = function(gd, data, layout, config) {
         gd.layout = layout || {};
         helpers.cleanLayout(gd.layout);
 
-        Plots.supplyDefaults(gd);
+        // "true" skips updating calcdata and remapping arrays from calcTransforms,
+        // which supplyDefaults usually does at the end, but we may need to NOT do
+        // if the diff (which we haven't determined yet) says we'll recalc
+        Plots.supplyDefaults(gd, true);
```

Comment:
I'm not a big fan of pattern. 

At the very least we should turn this into `Plots.supplyDefaults(gd, {skipDefaultsUpdateCalc: 1})` to make it more readable and extendable. 

Personally, I'd vote for removing the newly added `Plots.supplyDefaultsUpdateCalc()` from `Plots.supplyDefaults` and call `Plotly.supplyDefaultsUpdateCalc` after `Plots.supplyDefaults` in the places that new it.

On `master`, we have:

![image](https://user-images.githubusercontent.com/6675409/39137509-c3c0d778-46eb-11e8-9037-6ca5c7239397.png)

where the calls in `validate.js` and `rangeslider/draw.js` (and maybe more) don't need to update calc.

Reply:
cleaner `supplyDefaults` API -> https://github.com/plotly/plotly.js/pull/2577/commits/2a41f9ed9184014a543004ae6a5042334ccf8fec